tileview/src/main/java/com/qozix/tileview/tiles/TileRenderPoolExecutor.java

```diff
@@ -0,0 +1,115 @@
+    sInstance = new TileRenderPoolExecutor();
+  }
+
+  public TileRenderPoolExecutor() {
+    mQueue = new LinkedBlockingDeque<>();
+    mExecutor = new CustomThreadPoolExecutor( CORE_POOL_SIZE, MAXIMUM_POOL_SIZE, KEEP_ALIVE_TIME, KEEP_ALIVE_TIME_UNIT, mQueue );
+    mFutureList = new LinkedBlockingDeque<>();
+  }
+
+  public static TileRenderPoolExecutor getsInstance() {
+    return sInstance;
+  }
+
+  public void cancel() {
+    synchronized ( this ) {
+      for ( Future f : mFutureList ) {
+        if( !f.isDone() ){
+          f.cancel( true );
+        }
+      }
+      mQueue.clear();
+      mFutureList.clear();
+    }
+  }
+
+  public void queue( TileCanvasViewGroup viewGroup, LinkedList<Tile> tiles ) {
+    if( tiles != null && tiles.size() > 0) {
+      mViewGroup = new WeakReference<>( viewGroup );
+      viewGroup.onRenderTaskPreExecute();
+      int size = tiles.size();
+      for( int i = 0 ; i < size ; i++ ) {
+        mFutureList.add( mExecutor.submit( new TileRenderRunnable( viewGroup, tiles.get( i ) ) ) );
+      }
+    }
+  }
+
+  static class TileRenderRunnable implements Runnable {
+    private final WeakReference<TileCanvasViewGroup> mTileCanvasViewGroup;
+    private final WeakReference<Tile> mTile;
+
+    public TileRenderRunnable( TileCanvasViewGroup viewGroup, Tile tile) {
+      this.mTileCanvasViewGroup = new WeakReference<>( viewGroup );
+      this.mTile = new WeakReference<>( tile );
+    }
+
+    @Override
+    public void run() {
+      android.os.Process.setThreadPriority( android.os.Process.THREAD_PRIORITY_BACKGROUND );
+      final Thread thread = Thread.currentThread();
+
+      TileCanvasViewGroup tileCanvasViewGroup = mTileCanvasViewGroup.get();
+      Tile tile = mTile.get();
+
+      if( tileCanvasViewGroup != null ){
+        if( thread.isInterrupted() ){
+          tileCanvasViewGroup.onRenderTaskCancelled();
+          return;
+        }
+
+        if( !tileCanvasViewGroup.getRenderIsCancelled() && tile != null ){
+          tileCanvasViewGroup.generateTileBitmap( tile );
+          if ( thread.isInterrupted() ){
+            tileCanvasViewGroup.onRenderTaskCancelled();
+            return;
+          }
+
+          tileCanvasViewGroup.addTileToCurrentTileCanvasView( tile );
+        }
+      }
+    }
+  }
+
+  private class CustomThreadPoolExecutor extends ThreadPoolExecutor{
+    public CustomThreadPoolExecutor( int corePoolSize, int maximumPoolSize, long keepAliveTime, TimeUnit unit, BlockingQueue<Runnable> workQueue ) {
+      super( corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue );
+    }
+
+    @Override
+    protected void afterExecute( Runnable r, Throwable t ) {
+      super.afterExecute( r, t );
+//      getActiveCount() == 1 to make sure it is going to execute only for the last thread to run
+      if( mQueue != null && mQueue.size() == 0 && mViewGroup != null && mViewGroup.get() != null && getActiveCount() == 1) {
+        mViewGroup.get().onRenderTaskPostExecute();
```

Comment:
there is already a test for `mViewGroup != null && mViewGroup.get() != null`. Or is it about anything else?